dipy/denoise/localpca.py

```diff
@@ -145,16 +150,26 @@ def genpca(arr, sigma=None, mask=None, patch_radius=2, pca_method='eig',
         raise ValueError("patch_radius should have length 3")
     else:
         patch_radius = np.asarray(patch_radius).astype(int)
+    patch_radius[arr.shape[0:3] == np.ones(3)] = 0  # account for dim of size 1
     patch_size = 2 * patch_radius + 1
 
-    if np.prod(patch_size) < arr.shape[-1]:
-        e_s = "You asked for PCA denoising with a "
-        e_s += "patch_radius of {0} ".format(patch_radius)
-        e_s += "with total patch size of {0}".format(np.prod(patch_size))
-        e_s += "for data with {0} directions. ".format(arr.shape[-1])
-        e_s += "This would result in an ill-conditioned PCA matrix. "
-        e_s += "Please increase the patch_radius."
-        raise ValueError(e_s)
+    ash = arr.shape[0:3]
+    if np.any((ash != np.ones(3)) * (ash < patch_size)):
+        raise ValueError("Array 'arr' is incorrect shape")
+
+    num_samples = np.prod(patch_size)
+    if num_samples == 1:
+        raise ValueError("Cannot have only 1 sample, please increase patch_radius.")
+    if num_samples < arr.shape[-1]:
+        tmp = np.sum(patch_size == 1)  # count spatial dimensions with size 1
+        if tmp == 0: root = np.ceil(arr.shape[-1] ** (1./3))  # 3D
+        if tmp == 1: root = np.ceil(arr.shape[-1] ** (1./2))  # 2D
+        if tmp == 2: root = arr.shape[-1]  # 1D
+        root = root + 1 if (root % 2) == 0 else root  # make odd
+        spr = int((root - 1) / 2)  # suggested patch_radius
+        e_s = "Number of samples {1} < Dimensionality {0}. ".format(arr.shape[-1], num_samples)
+        e_s += "Consider increasing patch_radius to {0}.".format(spr)
+        warn(e_s, UserWarning)
```

Comment:
I still wonder whether we should raise an error. Beginner and intermediate users will experiment with patch radius without understanding the consequence. 

We prefer to stop the process, propose a solution, and let the user think.. If we allow the process to run a very long time, we might lose the user or the user come back to us with: "it does not work".

Reply:
I have included an argument "suppress_warning" to genpca(), which I have not used yet. It was place-holding while we decided what to do concerning this error or warning. We could change it to 'suppress_error' for example, so that only a warning is supplied if this is the case? That way, a user would be forced to consciously deactivate the error. However...

On the topic of whether this should be an error, I am not completely convinced that an error should be raised rather than a warning. From **Diffusion MRI noise mapping using random matrix theory: Diffusion MRI Noise Mapping**, with M number of voxels and N number of data components:

> However, one rather chooses M > N to avoid the situation when M becomes the dominating factor in the deviation from the MP law, which is of order O(1 / min(M, N))

This is the only mention in the paper about having more voxels than components. With P representing the number of components that can represent the bulk of the signal, the key thing is to ensure is that

> the measurements are highly redundant, P << N. First, small P can be achieved by limiting the diffusion-weighting. Indeed, the complexity of the underlying, generally non-Gaussian diffusion process, and as such P, tends to increase with the b-value. Second, P increases with the size M of the sliding window due to the potential heterogeneity in noise level within the set of voxels.

In the case that more voxels are required than signal directions, we may be introducing complexity due to including more variation of signals, also with different noise levels. I think this is likely to degrade performance much more than using M < N. It is often the case that many repeated acquisitions are performed, so many signals in N have been obtained with the same bvec and bval. This is unlike the case in the paper, which does not contain any repeated acquisitions. Repeated acquisitions will assist in denoising, but in theory they aren't adding much complexity to the data, so we would not expect P to grow much even if N were to grow massively (e.g. 10 repeated acquisitions per bvec and bval).

For these reasons, I believe that it is better to raise a warning, possibly allowing the user to suppress this warning if they find it irritating.

Reply:
Thanks @skoudoro I have done everything requested, but I have avoided using the term `performance degradation` in the warning - I am not sure that we can accurately make this claim. In my first comment in this pull request, I showed how M > N can be achieved simply by supplying copies of existing data, and yet this is not really adding new data. I can change this if it is a sticking point, but I think we would need to be cautious about our wording

Reply:
It would be great to add a word concerning performance. So you can use `it might have performance impact` or something similar.

Reply:
That's good wording, I will use that